### PR TITLE
[DS-96] Checkbox

### PR DESCRIPTION
**Pull request by victormath12**
This PR implements the Checkbox component on web and native platforms and some test configurations.

image

image

### CODE REVIEW

File: packages/yoga/src/Checkbox/native/Checkbox.jsx [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  const [pressed, setPressed] = useState(false);
+  return (
+    <>
+      <TouchableWithoutFeedback
+        accessibilityRole="checkbox"
+        onPress={() => !disabled && onChange()}
+        onPressIn={() => !disabled && setPressed(true)}
+        onPressOut={() => !disabled && setPressed(false)}
+        {...rest}
+      >
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/native/Checkbox.jsx]: Why not use the [withTouchable](https://github.com/Gympass/yoga/blob/master/packages/yoga/src/Button/native/withTouchable.jsx) HoC?

File: packages/yoga/src/Checkbox/native/Checkbox.jsx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  /** set a short helper text under checkbox */
+  helper: string,
+  /** set a checked state in the component */
+  checked: bool,
+  /** set a disabled state in the component */
+  disabled: bool,
+  /** set a error state in the component */
+  error: bool,
+  /** when the checkbox is checked/unchecked */
+  onChange: func,
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/native/Checkbox.jsx]: In my opinion, just the helper prop must receive its description. All the others are redundant and could be omitted. 

And, for all prop, you could avoid the prefix "set the|a"

File: packages/yoga/src/Checkbox/native/Checkbox.test.jsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
+import React from 'react';
+import { render, fireEvent, toJSON } from '@testing-library/react-native';
+
+import Checkbox from '..';
+import ThemeProvider from '../../ThemeProvider';
+
+const data = {
+  id: 'checkbox',
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/native/Checkbox.test.jsx]: This `id` is not required and could be omitted.

File: packages/yoga/src/Checkbox/native/Checkbox.test.jsx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+      const { container } = render(
+        <ThemeProvider>
+          <Checkbox {...data} error checked />
+        </ThemeProvider>,
+      );
+      expect(toJSON(container)).toMatchSnapshot();
+    });
+  });
+
+  describe('call events', () => {
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/native/Checkbox.test.jsx]:
```suggestion
  describe('Events', () => {
```

File: packages/yoga/src/Checkbox/web/CheckMark.jsx [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    theme: {
+      components: { checkbox },
+    },
+  }) => `
+    position: relative;
+    width: ${checkbox.width}px;
+    height: ${checkbox.height}px;
+    border-radius: ${checkbox.border.radii}px;
+    border-width: ${checkbox.border.width}px;
+    border-style: solid;
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/web/CheckMark.jsx]: `solid` is the default value of `border-style`. You could omit this value.

File: packages/yoga/src/Checkbox/web/CheckMark.jsx [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+    ${
+      checked
+        ? `background-color: ${checkbox.checked.backgroundColor}; 
+            background-image: url(${check}); 
+            background-size: 16px 16px;
+            background-repeat: no-repeat;
+            background-position: center;`
+        : ``
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/web/CheckMark.jsx]:
```suggestion
        : ''
```

File: packages/yoga/src/Checkbox/web/CheckMark.jsx [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    ${
+      disabled
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/web/CheckMark.jsx]: The same as in the native component

File: packages/yoga/src/Checkbox/web/CheckMark.jsx [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        ? `border-color: ${checkbox.disabled.border.color};`
+        : `border-color: ${checkbox.border.color};`
+    }
+
+    ${
+      disabled && checked
+        ? `background-color: ${checkbox.disabled.backgroundColor};`
+        : ``
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/web/CheckMark.jsx]:
```suggestion
        : ''
```

File: packages/yoga/src/Checkbox/web/CheckMark.jsx [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    ${error ? `border-color: ${checkbox.error.border.color};` : ``}
+
+    ${
+      error && checked
+        ? `background-color: ${checkbox.error.backgroundColor};`
+        : ``
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/web/CheckMark.jsx]:
```suggestion
        : ''
```

File: packages/yoga/src/Checkbox/web/CheckMark.jsx [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    }
+  `,
+);
+
+const CircleHover = styled.div(
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/web/CheckMark.jsx]: You don't need to make the same logic as you do in native. You could do it using `:before` here.

File: packages/yoga/src/Checkbox/web/CheckMark.jsx [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+  hovered,
+  disabled,
+  error,
+  onChange,
+  ...rest
+}) => (
+  <CheckMarkStyled checked={checked} disabled={disabled} error={error}>
+    <HiddenInput
+      type="checkbox"
+      value={value}
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/web/CheckMark.jsx]: consider omitting this prop

File: packages/yoga/src/Checkbox/web/Checkbox.jsx [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+      >
+        <CheckMark
+          id={id}
+          value={value}
+          hovered={hover}
+          disabled={disabled}
+          checked={checked}
+          error={error}
+        />
+        <Label htmlFor={id}>{label}</Label>
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/web/Checkbox.jsx]: Is the `htmlFor` really needed? Why not nesting CheckMark inside Label?

File: packages/yoga/src/Checkbox/web/Checkbox.jsx [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+  style,
+  onChange,
+  ...rest
+}) => {
+  const [hover, setHover] = useState(false);
+  return (
+    <CheckboxWrapper style={style}>
+      <CheckboxStyled
+        onClick={() => !disabled && onChange()}
+        onMouseEnter={() => !disabled && setHover(true)}
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/web/Checkbox.jsx]: Why hover is a state? 👀

File: packages/yoga/src/Checkbox/web/Checkbox.jsx [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+  /** set a checked state in the checkbox input */
+  checked: bool,
+  /** set a disabled state in the checkbox input */
+  disabled: bool,
+  /** set a error state in the component */
+  error: bool,
+  /** set a style to the checkbox container */
+  style: objectOf(any),
+  /** when the checkbox is checked/unchecked */
+  onChange: func,
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/web/Checkbox.jsx]: The same as in native

File: packages/yoga/src/Checkbox/web/Checkbox.test.jsx [hunk 15]
```diff
@@ -0,0 +1,83 @@
+import React from 'react';
+import { render, fireEvent } from '@testing-library/react';
+
+import Checkbox from '..';
+import ThemeProvider from '../../ThemeProvider';
+
+const data = {
+  id: 'checkbox',
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/web/Checkbox.test.jsx]: The `id` is not required in this component

File: packages/yoga/src/Checkbox/web/Checkbox.test.jsx [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+      const { container } = render(
+        <ThemeProvider>
+          <Checkbox {...data} error checked />
+        </ThemeProvider>,
+      );
+      expect(container).toMatchSnapshot();
+    });
+  });
+
+  describe('call events', () => {
```
Comment (ggdaltoso) [inline packages/yoga/src/Checkbox/web/Checkbox.test.jsx]:
```suggestion
  describe('Events', () => {
```

File: packages/yoga/src/ThemeProvider/themes/BaseTheme.js [hunk 17]
```diff
@@ -254,6 +256,53 @@ const BaseTheme = ({ primary, secondary }) => {
         color: colors.gray[2],
       },
     },
+    checkbox: {
+      width: spacing.xlarge,
+      height: spacing.xlarge,
+      border: {
+        width: borders.medium,
+        color: colors.primary[3],
+        radii: 4,
```
Comment (ggdaltoso) [inline packages/yoga/src/ThemeProvider/themes/BaseTheme.js]: ?

File: packages/yoga/src/Checkbox/web/CheckMark.jsx (hunk 5 above)
Comment (victormath12) [reply packages/yoga/src/Checkbox/web/CheckMark.jsx]: I have tested and this `border-style: solid` is necessary, isn't default.

File: packages/yoga/src/ThemeProvider/themes/BaseTheme.js (hunk 17 above)
Comment (victormath12) [reply packages/yoga/src/ThemeProvider/themes/BaseTheme.js]: We don't have a radii value `4` in our token, we just have `0`, `8` and `999`.
What do you suggest? Add radii `4` in our token file?

Comment (victormath12) [reply packages/yoga/src/ThemeProvider/themes/BaseTheme.js]: Another option is to use checkbox `border-radius: 4px` direct in the styled component without using token/theme variable.

Comment (ggdaltoso) [reply packages/yoga/src/ThemeProvider/themes/BaseTheme.js]: You should talk to the design guy and ask him about it.

File: packages/doc/content/components/checkbox/checkbox-web.mdx [hunk 18]
```diff
@@ -0,0 +1,29 @@
+### Reference
+
+Gympass `<Checkbox />` are used for a list of options where the user may select multiple options, including all or none.
+
+### Usage
+
+```javascript reactLive=true center=true
+<section>
+  <Checkbox
+    style={{ width: 200, marginBottom: 40 }}
```
[the remark this answers is not in the document]
Comment (victormath12) [reply packages/doc/content/components/checkbox/checkbox-web.mdx]: We decided to show the both cases (Checkbox checked and unchecked)

File: packages/yoga/src/Checkbox/native/Checkbox.jsx (hunk 1 above)
Comment (victormath12) [reply packages/yoga/src/Checkbox/native/Checkbox.jsx]: We have tested and in this case is not applicable.

File: packages/yoga/src/Checkbox/native/CheckMark.jsx [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+    },
+  }) => `
+    width: 24px;
+    height: 24px;
+
+    border-radius: ${checkbox.border.radii}px;
+    border-width: ${checkbox.border.width}px;
+    border-style: solid;
+
+    ${checked ? `background-color: ${checkbox.checked.backgroundColor};` : ''}
```
Comment (allyssonsantos) [inline packages/yoga/src/Checkbox/native/CheckMark.jsx]: In this case If the checkbox isn't checked the background will be transparent, is it correct?

File: packages/yoga/src/ThemeProvider/themes/BaseTheme.js (hunk 17 above)
Comment (victormath12) [reply packages/yoga/src/ThemeProvider/themes/BaseTheme.js]: We have talked with the DesignOps team and we decided to add a new `radii` token: `semiRouded`

File: packages/yoga/src/Checkbox/native/CheckMark.jsx (hunk 19 above)
Comment (victormath12) [reply packages/yoga/src/Checkbox/native/CheckMark.jsx]: In this moment we will not use a background or inverted state, maybe in a second moment we can implement this behavior

File: packages/yoga/src/Checkbox/native/CheckMark.jsx [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+        ? `background-color: ${checkbox.error.backgroundColor};`
+        : ''
+    }
+  `,
+);
+
+const CheckIcon = styled.Image`
+  width: 16px;
+  height: 11px;
+  margin: 4px 2px 4px 2px;
```
Comment (allyssonsantos) [inline packages/yoga/src/Checkbox/native/CheckMark.jsx]:
```suggestion
  margin: 4px 2px 4px;
```

File: packages/yoga/src/Checkbox/native/CheckMark.jsx [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+    left: -17px;
+
+    width: 54px;
+    height: 54px;
+    
+    background-color: ${hexToRgb(
+      checkbox.checked.backgroundColor,
+      0.2,
+    )};          
+    border-radius: ${checkbox.hover.border.radius};
```
Comment (allyssonsantos) [inline packages/yoga/src/Checkbox/native/CheckMark.jsx]:
```suggestion
    border-radius: ${checkbox.hover.border.radius}px;
```

File: packages/yoga/src/Checkbox/web/Checkbox.jsx [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+  `,
+);
+
+const HelperWrapper = styled.div(
+  ({
+    theme: {
+      components: { checkbox },
+    },
+  }) => `
+    display: flex;
```
Comment (allyssonsantos) [inline packages/yoga/src/Checkbox/web/Checkbox.jsx]: Why do you need this `display: flex`?

**Pull request merged**